### PR TITLE
column_copy: add support for weight vector

### PR DESCRIPTION
**Pull request by komainu8**
I fixed the result of `column_copy` is incorrect against Float32 values as below.

Problem:

Values of `destination_tags` column incorrect.

```
table_create Tags TABLE_HASH_KEY ShortText
[[0,0.0,0.0],true]
table_create CopyFloat32Value TABLE_HASH_KEY ShortText
[[0,0.0,0.0],true]
column_create CopyFloat32Value source_tags COLUMN_VECTOR|WITH_WEIGHT|WEIGHT_FLOAT32 Tags
[[0,0.0,0.0],true]
column_create CopyFloat32Value destination_tags COLUMN_VECTOR|WITH_WEIGHT|WEIGHT_FLOAT32 Tags
[[0,0.0,0.0],true]
load --table CopyFloat32Value
[
{
  "_key": "Groonga is fast!!!",
  "source_tags": {
    "Groonga": 2.8,
    "full text search": 1.5
  }
}
]
[[0,0.0,0.0],1]
column_copy CopyFloat32Value source_tags CopyFloat32Value destination_tags
[[0,0.0,0.0],true]
select CopyFloat32Value
[
  [
    0,
    0.0,
    0.0
  ],
  [
    [
      [
        1
      ],
      [
        [
          "_id",
          "UInt32"
        ],
        [
          "_key",
          "ShortText"
        ],
        [
          "destination_tags",
          "Tags"
        ],
        [
          "source_tags",
          "Tags"
        ]
      ],
      [
        1,
        "Groonga is fast!!!",
        {
          "Groonga": 2.802597e-45,
          "full text search": 0.0
        },
        {
          "Groonga": 2.8,
          "full text search": 1.5
        }
      ]
    ]
  ]
]
```

Expected:

Values of `destination_tags` column same as values of `source_tags` column as below.

```
column_copy CopyFloat32Value source_tags CopyFloat32Value destination_tags
[[0,0.0,0.0],true]
select CopyFloat32Value
[
  [
    0,
    0.0,
    0.0
  ],
  [
    [
      [
        1
      ],
      [
        [
          "_id",
          "UInt32"
        ],
        [
          "_key",
          "ShortText"
        ],
        [
          "destination_tags",
          "Tags"
        ],
        [
          "source_tags",
          "Tags"
        ]
      ],
      [
        1,
        "Groonga is fast!!!",
        {
          "Groonga": 2.8,
          "full text search": 1.5
        },
        {
          "Groonga": 2.8,
          "full text search": 1.5
        }
      ]
    ]
  ]
]
```

### CODE REVIEW

**Comment by komainu8** [comment]
@kou Could you review this PR?

File: lib/column.c [hunk 1]
```diff
@@ -266,8 +266,9 @@ grn_column_copy_same_table(grn_ctx *ctx,
     grn_obj *range = grn_ctx_at(ctx, to_range_id);
     GRN_TABLE_EACH_BEGIN(ctx, table, cursor, id) {
       GRN_BULK_REWIND(&value);
-      grn_obj_get_value(ctx, from_column, id, &value);
       GRN_BULK_REWIND(&validated_value);
+      grn_obj_get_value(ctx, from_column, id, &value);
+      grn_obj_get_value(ctx, from_column, id, &validated_value);
```
Comment (kou) [inline lib/column.c:271]: Why?

File: lib/column.c [hunk 2]
```diff
@@ -266,8 +266,9 @@ grn_column_copy_same_table(grn_ctx *ctx,
     grn_obj *range = grn_ctx_at(ctx, to_range_id);
     GRN_TABLE_EACH_BEGIN(ctx, table, cursor, id) {
       GRN_BULK_REWIND(&value);
-      grn_obj_get_value(ctx, from_column, id, &value);
       GRN_BULK_REWIND(&validated_value);
+      grn_obj_get_value(ctx, from_column, id, &value);
```
Comment (kou) [inline lib/column.c:270]: Why did you change the order?

**Comment by kou** [comment]
Does this work?

```diff
diff --git a/lib/column.c b/lib/column.c
index 32555a119..0f23f8fe8 100644
--- a/lib/column.c
+++ b/lib/column.c
@@ -262,13 +262,15 @@ grn_column_copy_same_table(grn_ctx *ctx,
   if (need_validate) {
     grn_obj validated_value;
     GRN_RECORD_INIT(&validated_value, to_range_flags, to_range_id);
+    if (to_range_flags & GRN_OBJ_WITH_WEIGHT) {
+      validated_value.header.flags |= GRN_OBJ_WITH_WEIGHT;
+    }
     bool is_uvector = grn_obj_is_uvector(ctx, &validated_value);
     grn_obj *range = grn_ctx_at(ctx, to_range_id);
     GRN_TABLE_EACH_BEGIN(ctx, table, cursor, id) {
       GRN_BULK_REWIND(&value);
-      GRN_BULK_REWIND(&validated_value);
       grn_obj_get_value(ctx, from_column, id, &value);
-      grn_obj_get_value(ctx, from_column, id, &validated_value);
+      GRN_BULK_REWIND(&validated_value);
       if (is_uvector) {
         uint32_t n_elements = grn_uvector_size(ctx, &value);
         uint32_t i;
```

File: lib/column.c (hunk 2 above)
Comment (komainu8) [reply lib/column.c:270]: Because I thought it would be easier to read if the same processes were grouped together.

File: lib/column.c (hunk 1 above)
Comment (kou) [reply lib/column.c:271]: `validated_value` has duplicated values because `grn_obj_get_value(ctx, &validated_value)` and `grn_uvector_add_element_record()` adds the same value twice.

Comment (komainu8) [reply lib/column.c:271]: Oh...
I see. I understand. Thank you for your comment.

**Comment by komainu8** [comment]
Thank you for your suggested code.
It works.

**Comment by kou** [comment]
Could you also add a test for only `WITH_WEIGHT` (no `WEIGHT_FLOAT32`)?

It seems that this is not related to `WEIGHT_FLOAT32`. This is just only related to `WITH_WEIGHT`.

**Comment by komainu8** [comment]
> Could you also add a test for only WITH_WEIGHT (no WEIGHT_FLOAT32)?
> It seems that this is not related to WEIGHT_FLOAT32. This is just only related to WITH_WEIGHT.

Thank you for your comment. I see.
I added the test at https://github.com/groonga/groonga/pull/1379/commits/846e6c3f4104e1b03aa7bd2efcd4d3f3e62eafce.

File: test/command/suite/column_copy/float32.test [hunk 3]
```diff
@@ -0,0 +1,19 @@
+table_create Tags TABLE_HASH_KEY ShortText
```
Comment (kou) [inline test/command/suite/column_copy/float32.test:1]: Could you rename this file to `with_float32.test`?

Comment (komainu8) [reply test/command/suite/column_copy/float32.test]: I see.
I fixed file name at https://github.com/groonga/groonga/pull/1379/commits/7b1ab445b36ec86ae2446fc2309e6f30d6137dbc.

Comment (kou) [reply test/command/suite/column_copy/float32.test]: Sorry. `weight_float32.test`...

Comment (komainu8) [reply test/command/suite/column_copy/float32.test]: Oh.
I fixed the file name again at https://github.com/groonga/groonga/pull/1379/commits/4da11d7f21459503c5e76076fd31d32d56d86ea5.
